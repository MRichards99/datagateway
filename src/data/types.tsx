export interface InvestigationData {
  ID: string;
  TITLE: string;
  VISIT_ID: number;
  RB_NUMBER: string;
  DOI: string;
  SIZE: number;
  INSTRUMENT: { NAME: string };
  STARTDATE: Date;
  ENDDATE: Date;
  [key: string]: string | number | Date | { NAME: string };
}

export interface DatasetData {
  ID: string;
  NAME: string;
  SIZE: number;
  CREATE_TIME: Date;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface DatafileData {
  ID: string;
  NAME: string;
  LOCATION: string;
  SIZE: number;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface FoodData {
  id: number;
  dessert: string;
  calories: number;
  fat: number;
  carbs: number;
  protein: number;
  [key: string]: string | number;
}

export type EntityType =
  | InvestigationData
  | DatasetData
  | DatafileData
  | FoodData;
