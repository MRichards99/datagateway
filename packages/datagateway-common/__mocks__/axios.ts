import axios from 'axios';

export default {
  get: jest.fn(() => Promise.resolve({ data: {} })),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};