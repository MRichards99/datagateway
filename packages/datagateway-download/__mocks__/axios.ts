import axios from 'axios';

export default {
  get: jest.fn(path => {
    if (path.includes('/topcat/user/cart/')) {
      return Promise.resolve({ data: { cartItems: [] } });
    } else if (path === '/datagateway-download-settings.json') {
      return Promise.resolve({
        data: {
          facilityName: 'LILS',
          apiUrl: 'http://scigateway-preprod.esc.rl.ac.uk:5000',
          downloadApiUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/topcat',
          idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
          accessMethods: {
            https: {
              idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
              displayName: 'HTTPS',
              description: 'Example description for HTTPS access method.',
            },
            globus: {
              idsUrl: 'https://scigateway-preprod.esc.rl.ac.uk:8181/ids',
              displayName: 'Globus',
              description: 'Example description for Globus access method.',
            },
          },
        },
      });
    } else {
      return Promise.resolve({ data: {} });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};
