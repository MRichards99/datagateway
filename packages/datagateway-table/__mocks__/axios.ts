import axios from 'axios';

export default {
  get: jest.fn(path => {
    if (path === '/settings.json') {
      return Promise.resolve({
        data: {
          'ui-strings': '/res/default.json',
        },
      });
    } else {
      return Promise.resolve({
        data: {},
      });
    }
  }),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};
