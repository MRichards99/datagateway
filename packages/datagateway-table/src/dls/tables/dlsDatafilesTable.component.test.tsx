import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import DLSDatafilesTable from './DLSDatafilesTable.component';
import { initialState } from '../../state/reducers/dgtable.reducer';
import configureStore from 'redux-mock-store';
import { StateType, Datafile } from '../../state/app.types';
import {
  fetchDatafilesRequest,
  filterTable,
  sortTable,
  downloadDatafileRequest,
} from '../../state/actions';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { TableSortLabel } from '@material-ui/core';
import Table from '../../table/table.component';
import { MemoryRouter } from 'react-router';
import axios from 'axios';

describe('DLS datafiles table component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;
  (axios.get as jest.Mock).mockImplementation(() =>
    Promise.resolve({ data: [] })
  );

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'div' });
    mount = createMount();

    mockStore = configureStore([thunk]);
    state = JSON.parse(JSON.stringify({ dgtable: initialState }));
    state.dgtable.data = [
      {
        ID: 1,
        NAME: 'Test 1',
        LOCATION: '/test1',
        FILESIZE: 1,
        CREATE_TIME: '2019-07-23',
        DATASET_ID: 1,
      },
    ];
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<DLSDatafilesTable store={mockStore(state)} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('sends fetchDatafiles action on load', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <DLSDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    expect(testStore.getActions()[0]).toEqual(fetchDatafilesRequest());
  });

  it('sends filterTable action on filter', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <DLSDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    const filterInput = wrapper.find('input').first();
    filterInput.instance().value = 'test';
    filterInput.simulate('change');

    expect(testStore.getActions()[1]).toEqual(filterTable('NAME', 'test'));

    filterInput.instance().value = '';
    filterInput.simulate('change');

    expect(testStore.getActions()[2]).toEqual(filterTable('NAME', null));
  });

  it('sends sortTable action on sort', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <DLSDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    wrapper
      .find(TableSortLabel)
      .first()
      .simulate('click');

    expect(testStore.getActions()[1]).toEqual(sortTable('NAME', 'asc'));
  });

  it('sends downloadData action on click of download button', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <DLSDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    wrapper.find('button[aria-label="Download"]').simulate('click');

    expect(testStore.getActions()[1]).toEqual(downloadDatafileRequest());
  });

  it("doesn't display download button for datafiles with no location", () => {
    const datafile = state.dgtable.data[0] as Datafile;
    const { LOCATION, ...datafileWithoutLocation } = datafile;
    state.dgtable.data = [datafileWithoutLocation];

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <MemoryRouter>
          <DLSDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('button[aria-label="Download"]')).toHaveLength(0);
  });

  it('renders details panel correctly', () => {
    const wrapper = shallow(
      <MemoryRouter>
        <DLSDatafilesTable store={mockStore(state)} datasetId="1" />
      </MemoryRouter>
    );
    const detailsPanelWrapper = shallow(
      wrapper.find(Table).prop('detailsPanel')(state.dgtable.data[0])
    );
    expect(detailsPanelWrapper).toMatchSnapshot();
  });

  it('renders file size as bytes', () => {
    const wrapper = mount(
      <Provider store={mockStore(state)}>
        <MemoryRouter>
          <DLSDatafilesTable datasetId="1" />
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper
        .find('[aria-colindex=4]')
        .find('p')
        .text()
    ).toEqual('1 B');
  });
});
