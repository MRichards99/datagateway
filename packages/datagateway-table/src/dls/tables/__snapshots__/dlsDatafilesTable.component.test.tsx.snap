// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLS datafiles table component renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "height": "calc(100vh - 64px)",
      "width": "100%",
    }
  }
>
  <WithStyles(VirtualizedTable)
    allIds={
      Array [
        1,
      ]
    }
    columns={
      Array [
        Object {
          "dataKey": "NAME",
          "filterComponent": [Function],
          "label": "Name",
        },
        Object {
          "dataKey": "LOCATION",
          "filterComponent": [Function],
          "label": "Location",
        },
        Object {
          "cellContentRenderer": [Function],
          "dataKey": "FILESIZE",
          "filterComponent": [Function],
          "label": "Size",
        },
        Object {
          "dataKey": "CREATE_TIME",
          "filterComponent": [Function],
          "label": "Create Time",
        },
      ]
    }
    data={
      Array [
        Object {
          "CREATE_TIME": "2019-07-23",
          "DATASET_ID": 1,
          "FILESIZE": 1,
          "ID": 1,
          "LOCATION": "/test1",
          "NAME": "Test 1",
        },
      ]
    }
    detailsPanel={[Function]}
    loadMoreRows={[Function]}
    loading={false}
    onCheck={[Function]}
    onSort={[Function]}
    onUncheck={[Function]}
    selectedRows={Array []}
    sort={Object {}}
    totalRowCount={0}
  />
</div>
`;

exports[`DLS datafiles table component renders details panel correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Typography))
    variant="body2"
  >
    <b>
      Name:
    </b>
     
    Test 1
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    variant="body2"
  >
    <b>
      Description:
    </b>
     
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    variant="body2"
  >
    <b>
      File Size:
    </b>
     
    1 B
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    variant="body2"
  >
    <b>
      Location:
    </b>
     
    /test1
  </WithStyles(ForwardRef(Typography))>
</div>
`;
